sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionProviderFromUri.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.rabbitmq;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.sdk.transforms.display.HasDisplayData;
+
+/**
+ * Modeled after {@link
+ * org.apache.beam.sdk.io.jdbc.JdbcIO.DataSourceProviderFromDataSourceConfiguration}, providing a
+ * means of obtaining a {@link ConnectionHandler} based on a static cache and a URI to promote
+ * re-use of existing connections rather than establishing a new connection for each RabbitMq
+ * interaction.
+ *
+ * <p>The cache is oriented around the full URI, which will not be parsed or validated until {@link
+ * #apply(Void)} is called.
+ *
+ * <p>Note that this can be shut down but it's unclear what aspect of the Beam runtime should be
+ * responsible for ensure this happens.
+ */
+public class ConnectionProviderFromUri
+    implements SerializableFunction<Void, ConnectionHandler>, HasDisplayData, Closeable {
+  private static final ConcurrentHashMap<String, ConnectionHandler> instances =
+      new ConcurrentHashMap<>();
+
+  private final String uri;
+  private final String displayableUri;
+
+  public ConnectionProviderFromUri(String uri) {
+    this.uri = uri;
+    String displayable = uri;
+    try {
+      URI parsed = URI.create(uri);
+      String userInfo =
+          Optional.ofNullable(parsed.getUserInfo())
+              .map(x -> x.split(":", 2)[0]) // user:pass
+              .map(username -> username + ":")
+              .orElse("");
```

Comment:
⛏ 
took me a sec to realize this is stripping the password.  might be straightforward to extract it to a function:
```java
      String userInfo =
          Optional.ofNullable(parsed.getUserInfo())
              .map(this::stripPassword)
              .orElse("");

...


  private String stripPassword(String userInfo) {
    return userInfo.split(":", 2)[0] + ":";
  }
```

Reply:
Yeah, this was surprisingly difficult to figure out how to do.